applications: nrf_desktop: Add click detector documentation (#1180)
NordicPlayground/fw-nrfconnect-nrf: applications/nrf_desktop/doc/click_detector.rst

.. _click_detector:

Click detector module
#####################

> The ``click_detector`` module is used to send ``click_event`` when recognized type of click is recorded for button defined in the module configuration.

Reviewer: ```suggestion
The ``click_detector`` module is used to send a ``click_event`` when a known type of click is recorded for the button defined in the module configuration.
```

Click type
**********

> The module records the following click types:

Reviewer: Rework the list to the following formatting:
* :cpp:enumerator:`CLICK_SHORT` - Button pressed and released after short time.

Author: done

* :cpp:enumerator:`CLICK_SHORT`
  Button pressed and released after short time.
* :cpp:enumerator:`CLICK_NONE`

>   Button pressed and held for time too long for :cpp:enumerator:`CLICK_SHORT`, but too short for :cpp:enumerator:`CLICK_LONG`.

Reviewer: ```suggestion
 Button pressed and held for a period of time that is too long for :cpp:enumerator:`CLICK_SHORT`, but too short for :cpp:enumerator:`CLICK_LONG`.
```

* :cpp:enumarator:`CLICK_LONG`

>   Button pressed and held for long time.

Reviewer: ```suggestion
 Button pressed and held for a long period of time.
```

* :cpp:enumerator:`CLICK_DOUBLE`

>   Two sequences of button press and release in short time interval.

Reviewer: ```suggestion
 Two sequences of the button press and release in a short time interval.
```

> Exact values of time intervals for click types are defined in ``click_detector.c`` file.

Reviewer: ```suggestion
The exact values of time intervals for click types are defined in the ``click_detector.c`` file.
```

Module Events
*************

+----------------+------------------+--------------------+------------------+------------------+
| Source Module  | Input Event      | This Module        | Output Event     | Sink Module      |
+================+==================+====================+==================+==================+
| :ref:`buttons` | ``button_event`` | ``click_detector`` | ``click_event``  | :ref:`ble_bond`  |
+----------------+------------------+--------------------+------------------+------------------+

Configuration
*************

Module detects click types based on ``button_event``.

> You should define :ref:`buttons` hardware interface.

Reviewer: ```suggestion
Make sure you define the :ref:`buttons` hardware interface.
```

> Set ``CONFIG_DESKTOP_CLICK_DETECTOR_ENABLE`` and define module configuration.

Reviewer: ```suggestion
Set ``CONFIG_DESKTOP_CLICK_DETECTOR_ENABLE`` and define the module configuration.
```

The configuration (array of :c:type:`struct click_detector_config`) is written in the ``click_detector_def.h`` file located in the board-specific directory in the application configuration folder.

> For every click detector you should define:

Reviewer: ```suggestion
For every click detector, you should define:
```

* :cpp:member:`key_id` - ID of selected key
* :cpp:member:`consume_button_event` - define if the ``button_event`` with given :cpp:member:`key_id` should be consumed by the module

Implementation details
**********************

Tracing of key states is implemented using periodically submitted work - :c:type:`struct k_delayed_work`.

> The work updates states of traced keys and sends ``click_event``.

Reviewer: ```suggestion
The work updates the states of traced keys and sends ``click_event``.
```

> The work is not submitted if there is no key, which state should be updated.

Reviewer: This is not clear. What is the meaning?

Author: Discussed offline
